File: swift/Sources/Shield/KeyRotationManager.swift
import Foundation
import CommonCrypto

/// KeyRotationManager - Version-based key management.
///
/// Supports seamless key rotation without breaking existing encrypted data.
/// Each ciphertext is tagged with the key version used.
///
/// Ciphertext format: version(4) || nonce(16) || ciphertext || mac(16)
public class KeyRotationManager {
    private static let nonceSize = 16
    private static let macSize = 16
    private static let minCiphertextSize = 4 + nonceSize + macSize

    private var keys: [Int: [UInt8]] = [:]
    private var _currentVersion: Int

    /// Get current key version.
    public var currentVersion: Int { _currentVersion }

    /// Get all available versions.
    public var versions: [Int] { keys.keys.sorted() }

    /// Create with initial key.
    public init(key: [UInt8], version: Int = 1) throws {
        guard key.count == 32 else {
            throw ShieldError.invalidKeySize
        }
        keys[version] = key
        _currentVersion = version
    }

    /// Add historical key for decryption.
    public func addKey(_ key: [UInt8], version: Int) throws {
        guard keys[version] == nil else {
            throw ShieldError.invalidToken
        }
        keys[version] = key
    }

    /// Rotate to new key.
    public func rotate(to newKey: [UInt8], version newVersion: Int? = nil) throws -> Int {
        let version = newVersion ?? _currentVersion + 1
        guard version > _currentVersion else {
            throw ShieldError.invalidToken
        }
        keys[version] = newKey
        _currentVersion = version
        return version
    }

    /// Encrypt with current key (includes version tag).
    public func encrypt(_ plaintext: [UInt8]) throws -> [UInt8] {
        guard let key = keys[_currentVersion] else {
            throw ShieldError.invalidKeySize
        }
        guard let nonce = Shield.randomBytes(KeyRotationManager.nonceSize) else {
            throw ShieldError.randomGenerationFailed
        }

        // Generate keystream and encrypt
        let keystream = generateKeystream(key: key, nonce: nonce, length: plaintext.count)
        var ciphertext = [UInt8](repeating: 0, count: plaintext.count)
        for i in 0..<plaintext.count {
            ciphertext[i] = plaintext[i] ^ keystream[i]
        }

        // Version bytes
        let versionBytes = withUnsafeBytes(of: UInt32(_currentVersion).littleEndian) { Array($0) }

        // HMAC authenticate (includes version)
        let macData = versionBytes + nonce + ciphertext
        let mac = Shield.hmacSha256(key: key, data: macData)

        // Result: version || nonce || ciphertext || mac
        return versionBytes + nonce + ciphertext + Array(mac.prefix(KeyRotationManager.macSize))
    }

    /// Decrypt with appropriate key version.
    public func decrypt(_ encrypted: [UInt8]) throws -> [UInt8] {
        guard encrypted.count >= KeyRotationManager.minCiphertextSize else {
            throw ShieldError.ciphertextTooShort
        }

        // Parse version
        let version = encrypted.withUnsafeBytes { buffer in
            buffer.load(as: UInt32.self).littleEndian
        }
        let nonce = Array(encrypted[4..<4 + KeyRotationManager.nonceSize])
        let ciphertext = Array(encrypted[4 + KeyRotationManager.nonceSize..<encrypted.count - KeyRotationManager.macSize])
        let receivedMac = Array(encrypted.suffix(KeyRotationManager.macSize))

        guard let key = keys[Int(version)] else {
            throw ShieldError.invalidToken
        }

        // Verify MAC
        let macData = Array(encrypted[0..<encrypted.count - KeyRotationManager.macSize])
        let expectedMac = Array(Shield.hmacSha256(key: key, data: macData).prefix(KeyRotationManager.macSize))

        guard Shield.constantTimeEquals(receivedMac, expectedMac) else {
            throw ShieldError.authenticationFailed
        }

        // Decrypt
        let keystream = generateKeystream(key: key, nonce: nonce, length: ciphertext.count)
        var plaintext = [UInt8](repeating: 0, count: ciphertext.count)
        for i in 0..<ciphertext.count {
            plaintext[i] = ciphertext[i] ^ keystream[i]
        }

        return plaintext
    }

    /// Re-encrypt data with current key.
    public func reEncrypt(_ encrypted: [UInt8]) throws -> [UInt8] {
        let plaintext = try decrypt(encrypted)
        return try encrypt(plaintext)
    }

    /// Remove old keys, keeping only recent versions.
    public func pruneOldKeys(keepVersions: Int = 2) -> [Int] {
        guard keepVersions >= 1 else { return [] }

        let sortedVersions = keys.keys.sorted().reversed()
        var toKeep = Set(sortedVersions.prefix(keepVersions))
        toKeep.insert(_currentVersion)

        var pruned: [Int] = []
        for v in keys.keys {
            if !toKeep.contains(v) {
                keys.removeValue(forKey: v)
                pruned.append(v)
            }
        }

        return pruned
    }

    /// Wipe all keys from memory.
    public func wipe() {
        for key in keys.keys {
            if var keyData = keys[key] {
                Shield.secureWipe(&keyData)
            }
        }
        keys.removeAll()
    }

    private func generateKeystream(key: [UInt8], nonce: [UInt8], length: Int) -> [UInt8] {
        let numBlocks = (length + 31) / 32
        var keystream = [UInt8]()
        keystream.reserveCapacity(numBlocks * 32)

        for i in 0..<numBlocks {
            var block = key + nonce
            block.append(contentsOf: withUnsafeBytes(of: UInt32(i).littleEndian) { Array($0) })
            let hash = Shield.sha256(block)
            keystream.append(contentsOf: hash)
        }

        return Array(keystream.prefix(length))
    }
}

File: csharp/Shield/KeyRotationManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;

namespace Guard8.Shield
{
    /// <summary>
    /// KeyRotationManager - Version-based key management.
    ///
    /// Supports seamless key rotation without breaking existing encrypted data.
    /// Each ciphertext is tagged with the key version used.
    ///
    /// Ciphertext format: version(4) || nonce(16) || ciphertext || mac(16)
    /// </summary>
    public class KeyRotationManager : IDisposable
    {
        private const int NonceSize = 16;
        private const int MacSize = 16;
        private const int MinCiphertextSize = 4 + NonceSize + MacSize;

        private readonly Dictionary<int, byte[]> _keys = new();
        private int _currentVersion;
        private bool _disposed;

        /// <summary>
        /// Create with initial key.
        /// </summary>
        public KeyRotationManager(byte[] key, int version = 1)
        {
            if (key.Length != 32)
                throw new ArgumentException("Key must be 32 bytes", nameof(key));

            var keyCopy = new byte[32];
            Array.Copy(key, keyCopy, 32);
            _keys[version] = keyCopy;
            _currentVersion = version;
        }

        /// <summary>
        /// Get current key version.
        /// </summary>
        public int CurrentVersion => _currentVersion;

        /// <summary>
        /// Get all available versions.
        /// </summary>
        public IEnumerable<int> Versions => _keys.Keys.OrderBy(v => v);

        /// <summary>
        /// Add historical key for decryption.
        /// </summary>
        public void AddKey(byte[] key, int version)
        {
            if (_keys.ContainsKey(version))
                throw new ArgumentException($"Version {version} already exists");

            var keyCopy = new byte[32];
            Array.Copy(key, keyCopy, 32);
            _keys[version] = keyCopy;
        }

        /// <summary>
        /// Rotate to new key.
        /// </summary>
        public int Rotate(byte[] newKey, int? newVersion = null)
        {
            int version = newVersion ?? _currentVersion + 1;
            if (version <= _currentVersion)
                throw new ArgumentException("New version must be greater than current");

            var keyCopy = new byte[32];
            Array.Copy(newKey, keyCopy, 32);
            _keys[version] = keyCopy;
            _currentVersion = version;
            return version;
        }

        /// <summary>
        /// Encrypt with current key (includes version tag).
        /// </summary>
        public byte[] Encrypt(byte[] plaintext)
        {
            byte[] key = _keys[_currentVersion];
            byte[] nonce = Shield.RandomBytes(NonceSize);

            // Generate keystream and encrypt
            byte[] keystream = GenerateKeystream(key, nonce, plaintext.Length);
            byte[] ciphertext = new byte[plaintext.Length];
            for (int i = 0; i < plaintext.Length; i++)
                ciphertext[i] = (byte)(plaintext[i] ^ keystream[i]);

            // Version bytes
            byte[] versionBytes = BitConverter.GetBytes(_currentVersion);

            // HMAC authenticate (includes version)
            byte[] macData = new byte[4 + NonceSize + ciphertext.Length];
            Array.Copy(versionBytes, 0, macData, 0, 4);
            Array.Copy(nonce, 0, macData, 4, NonceSize);
            Array.Copy(ciphertext, 0, macData, 4 + NonceSize, ciphertext.Length);
            byte[] mac = Shield.HmacSha256(key, macData);

            // Result: version || nonce || ciphertext || mac
            byte[] result = new byte[4 + NonceSize + ciphertext.Length + MacSize];
            Array.Copy(versionBytes, 0, result, 0, 4);
            Array.Copy(nonce, 0, result, 4, NonceSize);
            Array.Copy(ciphertext, 0, result, 4 + NonceSize, ciphertext.Length);
            Array.Copy(mac, 0, result, result.Length - MacSize, MacSize);

            return result;
        }

        /// <summary>
        /// Decrypt with appropriate key version.
        /// </summary>
        public byte[] Decrypt(byte[] encrypted)
        {
            if (encrypted.Length < MinCiphertextSize)
                throw new ArgumentException("Ciphertext too short");

            // Parse version
            int version = BitConverter.ToInt32(encrypted, 0);
            byte[] nonce = new byte[NonceSize];
            Array.Copy(encrypted, 4, nonce, 0, NonceSize);

            int ciphertextLen = encrypted.Length - 4 - NonceSize - MacSize;
            byte[] ciphertext = new byte[ciphertextLen];
            Array.Copy(encrypted, 4 + NonceSize, ciphertext, 0, ciphertextLen);

            byte[] receivedMac = new byte[MacSize];
            Array.Copy(encrypted, encrypted.Length - MacSize, receivedMac, 0, MacSize);

            if (!_keys.TryGetValue(version, out byte[] key))
                throw new ArgumentException($"Unknown key version: {version}");

            // Verify MAC
            byte[] macData = new byte[encrypted.Length - MacSize];
            Array.Copy(encrypted, 0, macData, 0, macData.Length);
            byte[] expectedMac = Shield.HmacSha256(key, macData);

            if (!Shield.ConstantTimeEquals(receivedMac, expectedMac, MacSize))
                throw new CryptographicException("Authentication failed");

            // Decrypt
            byte[] keystream = GenerateKeystream(key, nonce, ciphertextLen);
            byte[] plaintext = new byte[ciphertextLen];
            for (int i = 0; i < ciphertextLen; i++)
                plaintext[i] = (byte)(ciphertext[i] ^ keystream[i]);

            return plaintext;
        }

        /// <summary>
        /// Re-encrypt data with current key.
        /// </summary>
        public byte[] ReEncrypt(byte[] encrypted)
        {
            byte[] plaintext = Decrypt(encrypted);
            return Encrypt(plaintext);
        }

        /// <summary>
        /// Remove old keys, keeping only recent versions.
        /// </summary>
        public List<int> PruneOldKeys(int keepVersions = 2)
        {
            if (keepVersions < 1)
                throw new ArgumentException("Must keep at least 1 version");

            var versions = _keys.Keys.OrderByDescending(v => v).ToList();
            var toKeep = new HashSet<int>(versions.Take(keepVersions)) { _currentVersion };

            var pruned = new List<int>();
            foreach (var v in _keys.Keys.ToList())
            {
                if (!toKeep.Contains(v))
                {
                    Shield.SecureWipe(_keys[v]);
                    _keys.Remove(v);
                    pruned.Add(v);
                }
            }

            return pruned;
        }

        private static byte[] GenerateKeystream(byte[] key, byte[] nonce, int length)
        {
            int numBlocks = (length + 31) / 32;
            byte[] keystream = new byte[numBlocks * 32];

            for (int i = 0; i < numBlocks; i++)
            {
                byte[] block = new byte[32 + NonceSize + 4];
                Array.Copy(key, 0, block, 0, 32);
                Array.Copy(nonce, 0, block, 32, NonceSize);
                BitConverter.GetBytes(i).CopyTo(block, 32 + NonceSize);

                byte[] hash = Shield.Sha256(block);
                Array.Copy(hash, 0, keystream, i * 32, 32);
            }

            byte[] result = new byte[length];
            Array.Copy(keystream, result, length);
            return result;
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                foreach (var key in _keys.Values)
                    Shield.SecureWipe(key);
                _disposed = true;
            }
        }
    }
}
